### PR TITLE
feat: Implemented proper email-password authentication with conditional UI rendering.

### PR DESCRIPTION
**Pull request by Mohitsen11**
## Summary

This PR aims proper email-based authentication and protects routes which are only accessible for authenticated user. It does proper conditional rendering of UI and display appropriate success & errors notification when user performs actions.

## Description

This PR solves the auth problem and conditional UI rendering.

1. Application displays login button to unauthenticate user.
2. When user fill the signup/signin form data stores in DB and notify user success/error.
3. After successful authentication, application display create post + logout button.
4. Authenticate user can create post until access_token expires ( 2 min expiration time ) and created post ObjectID is added to user's createPosts field.
5. When token is expired and user try to create post, application notify error (with error-msg : Invalid user!) and redirects to homepage and display login button because now user has to login again.
6. Added icons for login and logout and aligned them correctly.

## Images
Screenshot 2024-04-22 163139

Screenshot 2024-04-22 163210

Screenshot 2024-04-22 163232

Screenshot 2024-04-22 163253

https://github.com/krishnaacharyaa/wanderlust/assets/114509045/8552a36f-476d-4ae5-95d5-e0b348875772

## Prerequisites

- [x] Have you followed all the [CONTRIBUTING GUIDELINES](https://github.com/krishnaacharyaa/wanderlust/blob/main/.github/CONTRIBUTING.md#guidelines-for-contributions)?

### CODE REVIEW

**Comment by krishnaacharyaa** [comment]
@Mohitsen11 Awesome work :) Loved the description of the PR. Thank you bunch for the efforts put in.

> When token is expired and user try to create post, application notify error (with error-msg : Invalid user!) and redirects to homepage and display login button because now user has to login again.

Can we may be say "Your session has expired, please login again!" instead of "Invalid User".Because user faced messages should be always friendly and normally shouldn't be like "Value null", "Invalid user" kinda thing, it should be like user understandable. The rest  code, I'll review once we are on the call and we go through the demo, I can't test because in the production, the environmental variable aren't there.

**Comment by krishnaacharyaa** [comment]
Hi @Mohitsen11, can you please let us know if you are busy, I'll try to pick this up... I'm aiming to finish this by today. Our test cases are being blocked because of this

**Comment by Mohitsen11** [comment]
> Hi @Mohitsen11, can you please let us know if you are busy, I'll try to pick this up... I'm aiming to finish this by today. Our test cases are being blocked because of this
> 

I will try to finish this today. And will take your help.

**Comment by krishnaacharyaa** [comment]
Yes please, let me know when we can sync up

**Comment by Mohitsen11** [comment]
> Yes please, let me know when we can sync up

Sure @krishnaacharyaa

File: backend/middlewares/auth-middleware.js
[the remark this answers is not in the document]
Comment (Mohitsen11) [reply backend/middlewares/auth-middleware.js:1]: Done.

File: backend/middlewares/auth-middleware.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { JWT_SECRET } from "../config/utils.js";
+import { HTTP_STATUS } from "../utils/constants.js";
+import jwt from 'jsonwebtoken';
+
+export const checkAuth = async (req, res, next) => {
+        const authHeader = req.headers.access_token;
```
[the remark this answers is not in the document]
Comment (Mohitsen11) [reply backend/middlewares/auth-middleware.js:6]: Done.

File: backend/middlewares/auth-middleware.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        if(authHeader){
+            const token = authHeader.split(" ")[1]; 
+
+            jwt.verify(token, JWT_SECRET, (error, payload) => {
+                if(error){
+                    return res.status(HTTP_STATUS.FORBIDDEN).json({
+                        success: false,
+                        message: "Token is not valid!",
+                    });
```
[the remark this answers is not in the document]
Comment (Mohitsen11) [reply backend/middlewares/auth-middleware.js:15]: Done.

File: frontend/src/layouts/header-layout.tsx [hunk 3]
```diff
@@ -1,10 +1,41 @@
...
+        setUser(null);
+        setIsLoggedIn(false);
+        localStorage.removeItem('isLoggedIn');
+        navigate('/');
+      } else {
+        toast.error('Error: ' + response.data.message);
+      }
+    } catch (err: any) {
+      toast.error('Error: ' + err.message);
+    }
```
[the remark this answers is not in the document]
Comment (Mohitsen11) [reply frontend/src/layouts/header-layout.tsx:37]: Done.

File: frontend/src/pages/add-blog.tsx [hunk 4]
```diff
@@ -106,7 +111,14 @@ function AddBlog() {
...
-        toast.error('Error: ' + err.message);
+        if (err.response.status === 403) {
+          toast.error('Error: ' + 'Your session has expired, please login again!');
+          setUser(null);
+          localStorage.setItem('isLoggedIn', 'false');
+          navigate('/');
+        } else {
+          toast.error('Error: ' + err.message);
+        }
       }
```
[the remark this answers is not in the document]
Comment (Mohitsen11) [reply frontend/src/pages/add-blog.tsx:122]: Done.

File: frontend/src/layouts/header-layout.tsx [hunk 5]
```diff
@@ -1,10 +1,41 @@
 import ThemeToggle from '@/components/theme-toggle-button';
 import AddIcon from '@/assets/svg/add-icon-white.svg';
+import { LogInIcon, LogOutIcon } from 'lucide-react';
 import { useNavigate } from 'react-router-dom';
```
[the remark this answers is not in the document]
Comment (Mohitsen11) [reply frontend/src/layouts/header-layout.tsx:4]: Done.

File: frontend/src/layouts/header-layout.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mohitsen11) [reply frontend/src/layouts/header-layout.tsx:37]: Done.

**Comment by Mohitsen11** [comment]
I made changes according to the review.

Now only localStorage + 3 Cookies is set make it 2  + delete contextProvider + The error messages should come in the input message error
Protected routes in the frontend should not be shown at all the add-blogs page are pending.

**Comment by Mohitsen11** [comment]
> I have made changes according to the review.
> 
> Now only localStorage + 3 Cookies is set make it 2 + delete contextProvider + The error messages should come in the input message error Protected routes in the frontend should not be shown at all the add-blogs page are pending.

Can you guide me how to use set-cookie in frontend returned in response by server after successful sigin/signup.

File: backend/.env.sample
[the remark this answers is not in the document]
Comment (Mohitsen11) [reply backend/.env.sample:1]: yes we can.

**Comment by krishnaacharyaa** [comment]
Sure @Mohitsen11 , https://www.youtube.com/watch?v=dX_LteE0NFM
check this video out, we'll do it this way, if you are facing issues let me know

**Comment by Mohitsen11** [comment]
> Sure @Mohitsen11 , https://www.youtube.com/watch?v=dX_LteE0NFM check this video out, we'll do it this way, if you are facing issues let me know

Cookies are not available in the application's cookies tab. When we had meet in discord we saw cookies were set there.

When I am logging or signing up, cookies are set by the server in the response headers.
In network tab set-cookie field is present with access_token and refresh_token in response headers but not available in application's cookie tab.

**Comment by krishnaacharyaa** [comment]
In this video, weren't we able to figure out ? The one shared above

**Comment by Mohitsen11** [comment]
> > Sure @Mohitsen11 , https://www.youtube.com/watch?v=dX_LteE0NFM check this video out, we'll do it this way, if you are facing issues let me know
> 
> Cookies are not available in the application's cookies tab. When we had meet in discord we saw cookies were set there.
> 
> When I am logging or signing up, cookies are set by the server in the response headers. In network tab set-cookie field is present with access_token and refresh_token in response headers but not available in application's cookie tab.

I got the solution.😊

**Comment by Mohitsen11** [comment]
> In this video, weren't we able to figure out ? The one shared above

Approach to remove localStorage :-
1. For removing the localStorage, we have to create an api which gets the token set by the server in browser and authenticate it and returns that token in response (if token is valid or not expired). And we'll make an api call from frontend ( basically from the home-layout page because where we want conditional rendering ) so that we can get the verified token and can set the isAuthenticated state and do the conditional rendering in UI. And we'll make api call inside useEffect hook with array containing single user state dependency. 

Any suggestions? @krishnaacharyaa

**Comment by krishnaacharyaa** [comment]
> Approach to remove localStorage :- For removing the localStorage, we have to create an api which gets the token set by the server in browser and authenticate it and returns that token in response (if token is valid or not expired). And we'll make an api call from frontend ( basically from the home-layout page because where we want conditional rendering ) so that we can get the verified token and can set the isAuthenticated state and do the conditional rendering in UI. And we'll make api call inside useEffect hook with array containing single user state dependency.

Why extra api call? Can't we leverage the cookie's variable directly by accessing it!

Let's have a call, and address all the issues...

**Comment by Mohitsen11** [comment]
> Owner

Sure.

File: frontend/src/App.tsx [hunk 6]
```diff
@@ -6,11 +6,17 @@ import ScrollToTop from '@/components/scroll-to-top';
...
+import { useState } from 'react';
 
 function App() {
+  const [user, setUser] = useState(null);
+
   return (
     <BrowserRouter>
       <ScrollToTop />
+      <UserContextProvider value={{ user, setUser } as UserContextType}>
       <div className="flex min-h-screen flex-col">
```
[the remark this answers is not in the document]
Comment (Mohitsen11) [reply frontend/src/App.tsx:20]: Done.

**Comment by Mohitsen11** [comment]
> > In this video, weren't we able to figure out ? The one shared above
> 
> 1. Approach to remove localStorage :-  For removing the localStorage, we have to create an api which gets the token set by the server in browser and authenticate it and returns that token in response (if token is valid or not expired). And we'll make an api call from frontend ( basically from the home-layout page because where we want conditional rendering ) so that we can get the verified token and can set the isAuthenticated state and do the conditional rendering in UI. And we'll make api call inside useEffect hook with array containing single user state dependency.
> 2. If we don't use context-provider then we'll have to leverage the above approach. after getting the token we'll send with request headers.
> 3. "The error messages should come in the input message error", I close to solve this.
> 4. If unauthenticate user try to access (hit)  /add-blog route, we'll simply redirect to a page that will display You are not authorized to access this page.
> 
> Any suggestions? @krishnaacharyaa

3rd task is done.

File: frontend/src/layouts/header-layout.tsx [hunk 7]
```diff
@@ -1,10 +1,40 @@
...
 import Hero from '@/components/hero';
+import { useEffect, useState } from 'react';
+import axios from 'axios';
+import { toast } from 'react-toastify';
 function header() {
   const navigate = useNavigate();
 
+  const [isLoggedIn, setIsLoggedIn] = useState(localStorage.getItem('isLoggedIn') === 'true');
+
```
Comment (krishnaacharyaa) [inline frontend/src/layouts/header-layout.tsx:14]: Have we not refactored this yet? is it a mistake?

File: frontend/src/layouts/header-layout.tsx [hunk 8]
```diff
@@ -13,28 +43,66 @@ function header() {
...
-              name="add-blog-mobile"
-              className="px-2 py-2 hover:bg-slate-500/25 md:hidden"
-              onClick={() => {
-                navigate('/add-blog');
-              }}
-            >
-              <img className="h-10 w-10" src={AddIcon} alt="add-blog" />
-            </button>
+            {isLoggedIn ? (
+              <div className="flex gap-2 ">
```
Comment (krishnaacharyaa) [inline frontend/src/layouts/header-layout.tsx:51]: And seems we are using it... in the frontend ...

**Review by krishnaacharyaa — COMMENTED** [review]
Hey @Mohitsen11, Great work, kudos to you for figuring out multiple things and doing a great contribution. Thanks a bunch.

Just a couple of clarification on the localstorage part and then we are very good to merge this.

**Comment by Mohitsen11** [comment]
> Hey @Mohitsen11, Great work, kudos to you for figuring out multiple things and doing a great contribution. Thanks a bunch.
> 
> Just a couple of clarification on the localstorage part and then we are very good to merge this.

I made the changes. 
 
Only persistent state management is pending and overall work is done.

And one more thing, when user clicks on logout button cookie must remove from browser's cookie but it is not happening.
I checked out the /signout api change the maxAge from 1 to 0. but cookie is still present in browser's cookie.

They remove when they are expired and page refreshes.

**Review by krishnaacharyaa — APPROVED** [review]
Looks great!, Singleton especially loved the way you took it up and implemented. 
Kudos to us, thanks to you :)

Merging it now

**Pull request merged**